### PR TITLE
[TM_TENSOR] Add tm_tensor.sort op and add it's lowering to Linalg_Ext

### PR DESCRIPTION
**Pull request by Abhishek-Varma**
-- This commit adds [tm_tensor.sort](https://github.com/llvm/torch-mlir/pull/2003) op and adds its lowering to
   iree_linalg_ext.sort op.

Signed-off-by: Abhishek Varma <abhishek@nod-labs.com>

### CODE REVIEW

**Comment by GMNGeoffrey** [comment]
So, when TM-Tensor was copied into IREE [almost a year ago](https://github.com/openxla/iree/commit/c39d296feb037f4348c1b407b9369a941ce8d1d4) now it was with assurances that it would be going upstream in "a matter of weeks". I think we need to reconsider where the right home for this is. I don't want to block this particular PR on that issue, but it brought it to mind again. @jpienaar @silvasean. Happy to split the discussion off into a separate issue.

**Comment by Abhishek-Varma** [comment]
Hi can someone review this?

**Comment by GMNGeoffrey** [comment]
> Hi can someone review this?

Likely people are holding off due to the CLA issue above. You will need to sign the CLA (or Nod will need to switch to a corporate CLA) before we can accept this contribution

**Comment by Abhishek-Varma** [comment]
> > Hi can someone review this?
> 
> Likely people are holding off due to the CLA issue above. You will need to sign the CLA (or Nod will need to switch to a corporate CLA) before we can accept this contribution

Hi @GMNGeoffrey,

Thanks for pointing it out!
I've signed up the CLA and there shouldn't be any issue on that front anymore.
Although I rebased + force-pushed to initiate CI action but to no avail.

**Comment by powderluv** [comment]
Going to approve and run the CI here since it looks like that is required for a first time contribution

**Comment by powderluv** [comment]
@jpienaar / @MaheshRavishankar can you please review / approve ?

**Comment by vivekkhandelwal1** [comment]
Hi @jpienaar @MaheshRavishankar, I also need this patch to run a model through Torch-MLIR + IREE path. Can we please get this approved/merged?

CC: @powderluv

File: third_party/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/TMTensor/IR/TMTensorOps.td [hunk 1]
```diff
@@ -189,6 +189,56 @@ def TMTensor_ScatterOp : TMTensor_Op<"scatter",
   }];
 }
 
+def TMTensor_SortOp : TMTensor_Op<"sort",
+    [DeclareOpInterfaceMethods<ScalarLoopOpInterface,
+        ["generateScalarImplementation"]>]> {
+  let summary = "Sort operator";
+  let description = [{
+    Based on XLA operation semantics, sorts the given `operands` at the given
```
Comment (jpienaar) [inline third_party/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/TMTensor/IR/TMTensorOps.td:197]: This is a little bit weird: is this a tm tensor one as where you are producing this you can't use StableHLO? Else why not use that StableHLO op? (If it is a case about mix of ops not being supported, then that's also something we should look at and has been on my list but not many requests).

File: third_party/torch-mlir-dialects/lib/Dialect/TMTensor/IR/TMTensorOps.cpp [hunk 2]
```diff
@@ -438,6 +438,166 @@ LogicalResult ScatterOp::generateScalarImplementation(OpBuilder &b,
...
+               << argType;
+      }
+    }
+  }
+
+  auto yieldOp = cast<YieldOp>(block.getTerminator());
+  if (yieldOp.getNumOperands() != 1) {
+    return op->emitOpError("should yield exactly one operand");
+  }
+  auto ty = yieldOp.getOperand(0).getType().dyn_cast<IntegerType>();
```
Comment (jpienaar) [inline third_party/torch-mlir-dialects/lib/Dialect/TMTensor/IR/TMTensorOps.cpp]: Nit: new MLIR style is dyn_cast<A>(b) vs b.dyn_cast<A>() that was previously required.

File: third_party/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/TMTensor/IR/TMTensorOps.td (hunk 1 above)
Comment (Abhishek-Varma) [reply third_party/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/TMTensor/IR/TMTensorOps.td:197]: So, this is needed in the lowering path from `torch-mlir` -> `linalg`.
We don't use `StableHLO` in that path, so we'd be requiring to use this TMTensor op. Here's the corresponding [Torch-MLIR to TMTensor SortOp](https://github.com/llvm/torch-mlir/pull/2003) PR, that'll need the current PR to go in.

CC: @powderluv

File: third_party/torch-mlir-dialects/lib/Dialect/TMTensor/IR/TMTensorOps.cpp [hunk 3]
```diff
@@ -438,6 +438,166 @@ LogicalResult ScatterOp::generateScalarImplementation(OpBuilder &b,
...
 }
 
+//===----------------------------------------------------------------------===//
+// SortOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult SortOp::verify() {
+  Operation *op = getOperation();
+  if (getNumInputs()) {
+    return op->emitOpError("does not expect to take any inputs");
```
Comment (jpienaar) [inline third_party/torch-mlir-dialects/lib/Dialect/TMTensor/IR/TMTensorOps.cpp]: This should be autogenerated for you

**Review by jpienaar — COMMENTED** [review]
This mostly looks reasonable to me, bubblesort feels a bit off but I'm not sure if there is some higherlevel part/this is a fallback.

We need to find final state for these, but I don't intend to block here on that.

But I'd like at least someone from codegen side to look re expanding tm_tensor before we get to better state.

**Comment by Abhishek-Varma** [comment]
Hi @jpienaar ,

Can this be approved and merged if there are no other review comments?

**Comment by GMNGeoffrey** [comment]
> Can this be approved and merged if there are no other review comments?

I think the missing bit was

> I'd like at least someone from codegen side to look re expanding tm_tensor before we get to better state.

@MaheshRavishankar can you take a look or delegate?

File: compiler/src/iree/compiler/InputConversion/TMTensor/test/convert_tm_tensor_to_linalg_ext.mlir [hunk 4]
```diff
@@ -76,3 +76,23 @@ func.func @scatter_add(
...
+// CHECK-LABEL:   func.func @sort(
+// CHECK-SAME:            %[[INPUT:.*]]: tensor<3x4xf32>, %[[INDICES:.*]]: tensor<3x4xi64>) -> (tensor<3x4xf32>, tensor<3x4xi64>) {
+// CHECK:          %[[SORT:.*]]:2 = iree_linalg_ext.sort
+// CHECK-SAME:            dimension(1)
+// CHECK-SAME:            outs(%[[INPUT]], %[[INDICES]] : tensor<3x4xf32>, tensor<3x4xi64>) {
+// CHECK:          ^bb0(%[[INP1:.*]]: f32, %[[INP2:.*]]: f32, %{{.*}}: i64, %{{.*}}: i64):
+// CHECK:            %[[PREDICATE:.*]] = arith.cmpf ole, %[[INP1]], %[[INP2]] : f32
+// CHECK:            iree_linalg_ext.yield %[[PREDICATE]] : i1
+// CHECK:          } -> tensor<3x4xf32>, tensor<3x4xi64>
+// CHECK:          return %[[SORT]]#0, %[[SORT]]#1 : tensor<3x4xf32>, tensor<3x4xi64>
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/InputConversion/TMTensor/test/convert_tm_tensor_to_linalg_ext.mlir]: Nit : newline.

**Review by MaheshRavishankar — APPROVED** [review]

**Comment by Abhishek-Varma** [comment]
Hi @MaheshRavishankar,

I've addressed your review comment. This is now ready to be merged.

**Comment by MaheshRavishankar** [comment]
Sorry, didnt realize that I had to approve... Ill let CI run and then merge...

**Comment by Abhishek-Varma** [comment]
> Sorry, didnt realize that I had to approve... Ill let CI run and then merge...

No worries. Thank you @MaheshRavishankar - we can get this in now.

**Pull request merged**